[macOS] Add style adjustment methods and paint method for color swatch for vector-based controls

lilyspiniolas opened the pull request:
#### 8801f12419402594b4b2b66a282f75f7a36507d7
<pre>
[macOS] Add style adjustment methods and paint method for color swatch for vector-based controls
<a href="https://bugs.webkit.org/show_bug.cgi?id=290776#">https://bugs.webkit.org/show_bug.cgi?id=290776#</a>
<a href="https://rdar.apple.com/148261180">rdar://148261180</a>

Reviewed by Aditya Keerthi.

Updated support for the vector-based color control. Added new StyleAppearances
`ColorWellSwatch`, `ColorWellSwatchOverlay`, and `ColorWellSwatchWrapper`
so that style adjustments can be performed on the input&apos;s shadow DOM elements
as needed, and so that the color swatch can be painted if needed.

Updated Mac test expectations for fast/forms/color/color-input-swatch.html
to include the new user agent part.

Added two new layout tests. `input-color-swatch-overlay-appearance.html`
ensures that, when the color swatch has `appearance: none`, the color
swatch overlay has a used appearance of &apos;none&apos;. This is done by inferring
the used appearance based on the overlay&apos;s computed styles.

`input-color-parts-appearance.html` ensures that a color input with
style `appearance: none` displays the same as a color input with
`appearance: none` set for the input, its color swatch wrapper,
and its color swatch.

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/color/input-color-parts-appearance-expected.html: Added.
* LayoutTests/fast/forms/color/input-color-parts-appearance.html: Added.
* LayoutTests/fast/forms/color/input-color-swatch-overlay-appearance-expected.txt: Added.
* LayoutTests/fast/forms/color/input-color-swatch-overlay-appearance.html: Added.
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac/fast/forms/color/color-input-swatch-expected.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/html.css:
(input[type=&quot;color&quot;]::-internal-color-swatch-overlay):
(input[type=&quot;color&quot;]::-webkit-color-swatch-wrapper):
(input[type=&quot;color&quot;]::-webkit-color-swatch):
* Source/WebCore/platform/StyleAppearance.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/StyleAppearance.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::parentOfElementUsesPrimitiveAppearance):
(WebCore::RenderTheme::adjustAppearanceForElement const):
(WebCore::RenderTheme::adjustStyle):
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::paint):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::adjustColorWellSwatchStyle const):
(WebCore::RenderTheme::adjustColorWellSwatchOverlayStyle const):
(WebCore::RenderTheme::adjustColorWellSwatchWrapperStyle const):
(WebCore::RenderTheme::paintColorWellSwatch):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::adjustColorWellSwatchStyle const):
(WebCore::RenderThemeCocoa::adjustColorWellSwatchOverlayStyle const):
(WebCore::RenderThemeCocoa::adjustColorWellSwatchWrapperStyle const):
(WebCore::RenderThemeCocoa::paintColorWellSwatch):
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::createColorWellSwatchSubtree):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/293442@main">https://commits.webkit.org/293442@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ee4d743c769cdfd7c8adaef80c4a5bf321f29a1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18461 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8700 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103950 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49413 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100869 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26911 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75233 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32369 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101828 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14250 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89251 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55593 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14042 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7223 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48793 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83983 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7294 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106319 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25921 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18896 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84197 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26296 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85449 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83695 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21239 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28349 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6019 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19632 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25874 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31060 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25694 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29014 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27268 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/css/html.css:
```diff
@@ -1082,6 +1083,7 @@ input[type="color"]::-webkit-color-swatch-wrapper {
 }
 
 input[type="color"]::-webkit-color-swatch {
+    appearance:auto;
```
```suggestion
    appearance: auto;
```

pxlcoder commented on Source/WebCore/css/html.css:
```diff
@@ -1068,6 +1068,7 @@ input[type="color"] {
 }
 
 input[type="color"]::-webkit-color-swatch-wrapper {
+    appearance:auto;
```
```suggestion
    appearance: auto;
```

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -98,13 +98,39 @@ RenderTheme::RenderTheme()
+static bool parentOfElementUsesPrimitiveAppearance(const Element* element)
+{
+    if (RefPtr parent = element->parentOrShadowHostElement()) {
+        const auto computedStyle = parent->computedStyle();
```
```suggestion
        CheckedPtr computedStyle = parent->computedStyle();
```

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -98,13 +98,39 @@ RenderTheme::RenderTheme()
+{
+    if (RefPtr parent = element->parentOrShadowHostElement()) {
+        const auto computedStyle = parent->computedStyle();
+        return computedStyle->usedAppearance() == StyleAppearance::None || computedStyle->nativeAppearanceDisabled();
```
Do we really need to check both cases? Since `nativeAppearanceDisabled` should set the used appearance to none after style adjustment.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -98,13 +98,39 @@ RenderTheme::RenderTheme()
 
 RenderTheme::~RenderTheme() = default;
 
+static bool parentOfElementUsesPrimitiveAppearance(const Element* element)
```
This can take a reference parameter.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -98,13 +98,39 @@ RenderTheme::RenderTheme()
 
+static bool parentOfElementUsesPrimitiveAppearance(const Element* element)
+{
+    if (RefPtr parent = element->parentOrShadowHostElement()) {
```
Shouldn't we always use the shadow host node?

lilyspiniolas replied:
It is possible that the color swatch has appearance:none, but the shadow host node doesn't. In this case, the overlay should have appearance:none too, so it has to check it's parent rather than the shadow host

pxlcoder replied:
`parentOrShadowHostElement` always returns the shadow host if one exists though:

```
inline ContainerNode* Node::parentOrShadowHostNode() const
{
    ASSERT(isMainThreadOrGCThread());
    if (auto* shadowRoot = dynamicDowncast<ShadowRoot>(*this))
        return shadowRoot->host();
    return parentNode();
}
```

pxlcoder replied:
(I also think we should add some new layout tests for this behavior)

lilyspiniolas replied:
Maybe I'm misunderstanding, but based on the code in `parentOrShadowHostElement` and `parentOrShadowHostNode` it seems like it only returns a shadow host node when the element's parent or the element itself is a shadow root, and returns a parent node otherwise. If it always returned the shadow host I would expect the behavior to be the same when using `shadowHost()` over `parentOrShadowHost()`, but it only behaves correctly when using the latter.

pxlcoder replied:
You're right, I missed the rest of the logic in `parentOrShadowHostElement()` 🤦 

(still think we should add some tests though)

pxlcoder reviewed (approved):
The tests should use spaces rather than tabs 🙂

lilyspiniolas committed:
[macOS] Add style adjustment methods and paint method for color swatch for vector-based controls
https://bugs.webkit.org/show_bug.cgi?id=290776#
rdar://148261180

Reviewed by Aditya Keerthi.

Updated support for the vector-based color control. Added new StyleAppearances
`ColorWellSwatch`, `ColorWellSwatchOverlay`, and `ColorWellSwatchWrapper`
so that style adjustments can be performed on the input's shadow DOM elements
as needed, and so that the color swatch can be painted if needed.

Updated Mac test expectations for fast/forms/color/color-input-swatch.html
to include the new user agent part.

Added two new layout tests. `input-color-swatch-overlay-appearance.html`
ensures that, when the color swatch has `appearance: none`, the color
swatch overlay has a used appearance of 'none'. This is done by inferring
the used appearance based on the overlay's computed styles.

`input-color-parts-appearance.html` ensures that a color input with
style `appearance: none` displays the same as a color input with
`appearance: none` set for the input, its color swatch wrapper,
and its color swatch.

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/color/input-color-parts-appearance-expected.html: Added.
* LayoutTests/fast/forms/color/input-color-parts-appearance.html: Added.
* LayoutTests/fast/forms/color/input-color-swatch-overlay-appearance-expected.txt: Added.
* LayoutTests/fast/forms/color/input-color-swatch-overlay-appearance.html: Added.
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac/fast/forms/color/color-input-swatch-expected.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/html.css:
(input[type="color"]::-internal-color-swatch-overlay):
(input[type="color"]::-webkit-color-swatch-wrapper):
(input[type="color"]::-webkit-color-swatch):
* Source/WebCore/platform/StyleAppearance.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/StyleAppearance.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::parentOfElementUsesPrimitiveAppearance):
(WebCore::RenderTheme::adjustAppearanceForElement const):
(WebCore::RenderTheme::adjustStyle):
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::paint):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::adjustColorWellSwatchStyle const):
(WebCore::RenderTheme::adjustColorWellSwatchOverlayStyle const):
(WebCore::RenderTheme::adjustColorWellSwatchWrapperStyle const):
(WebCore::RenderTheme::paintColorWellSwatch):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::adjustColorWellSwatchStyle const):
(WebCore::RenderThemeCocoa::adjustColorWellSwatchOverlayStyle const):
(WebCore::RenderThemeCocoa::adjustColorWellSwatchWrapperStyle const):
(WebCore::RenderThemeCocoa::paintColorWellSwatch):
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::createColorWellSwatchSubtree):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/293442@main
